### ISSUE TITLE
`pnpm run dev` fails in backend component

### ISSUE THREAD

**Issue by AndreasNasman** [ty-ras/meta#12]
There seems to be some type package that can't be found:

```sh
❯ pnpm run dev

> @my-project/backend@1.0.0 dev /home/nasse/tyras/my-project/components/backend
> MY_BACKEND_CONFIG=./config-dev.json pnpm run server-ts


> @my-project/backend@1.0.0 server-ts /home/nasse/tyras/my-project/components/backend
> ts-node src/index

/home/nasse/tyras/my-project/node_modules/.pnpm/ts-node@10.9.1_@types+node@18.16.5_typescript@5.0.4/node_modules/ts-node/src/index.ts:859
    return new TSError(diagnosticText, diagnosticCodes, diagnostics);
           ^
TSError: ⨯ Unable to compile TypeScript:
src/api/endpoints/openapi.ts:2:38 - error TS2307: Cannot find module 'openapi-types' or its corresponding type declarations.

2 import { OpenAPIV3 as openapi } from "openapi-types";
                                       ~~~~~~~~~~~~~~~

    at createTSError (/home/nasse/tyras/my-project/node_modules/.pnpm/ts-node@10.9.1_@types+node@18.16.5_typescript@5.0.4/node_modules/ts-node/src/index.ts:859:12)
    at reportTSError (/home/nasse/tyras/my-project/node_modules/.pnpm/ts-node@10.9.1_@types+node@18.16.5_typescript@5.0.4/node_modules/ts-node/src/index.ts:863:19)
    at getOutput (/home/nasse/tyras/my-project/node_modules/.pnpm/ts-node@10.9.1_@types+node@18.16.5_typescript@5.0.4/node_modules/ts-node/src/index.ts:1077:36)
    at Object.compile (/home/nasse/tyras/my-project/node_modules/.pnpm/ts-node@10.9.1_@types+node@18.16.5_typescript@5.0.4/node_modules/ts-node/src/index.ts:1433:41)
    at transformSource (/home/nasse/tyras/my-project/node_modules/.pnpm/ts-node@10.9.1_@types+node@18.16.5_typescript@5.0.4/node_modules/ts-node/src/esm.ts:400:37)
    at /home/nasse/tyras/my-project/node_modules/.pnpm/ts-node@10.9.1_@types+node@18.16.5_typescript@5.0.4/node_modules/ts-node/src/esm.ts:278:53
    at async addShortCircuitFlag (/home/nasse/tyras/my-project/node_modules/.pnpm/ts-node@10.9.1_@types+node@18.16.5_typescript@5.0.4/node_modules/ts-node/src/esm.ts:409:15)
    at async nextLoad (node:internal/modules/esm/loader:163:22)
    at async ESMLoader.load (node:internal/modules/esm/loader:605:20)
    at async ESMLoader.moduleProvider (node:internal/modules/esm/loader:457:11) {
  diagnosticCodes: [ 2307 ]
}
 ELIFECYCLE  Command failed with exit code 1.
 ELIFECYCLE  Command failed with exit code 1.
```

**Comment by AndreasNasman** [comment]
I'll investigate if I can find the source of the problem. I've had this issue before where packages aren't correctly structured and actually have missing dependencies but still work due to structure "flaws" in npm and Yarn 😅

**Comment by stazz** [comment]
This has to be something `pnpm`-specific. The `openapi-types` is [dependency of `@ty-ras/backend-node-io-ts-openapi`](https://github.com/ty-ras/packaging/blob/bb5a40576becfcba71a07c5b5fcc0c712f12095a/backend-node-io-ts-openapi/package.json#L56) and [also `@ty-ras/backend-node-zod-openapi`](https://github.com/ty-ras/packaging/blob/bb5a40576becfcba71a07c5b5fcc0c712f12095a/backend-node-zod-openapi/package.json#L53)  packages.

If you could dig some more info about this, I would be grateful. :)

**Comment by AndreasNasman** [comment]
pnpm also requires all dependencies to be explicitly listed in `package.json` for sanity which could also be the root cause here 👍

**Comment by stazz** [comment]
Could this be because sub-projects of pnpm-workspace (I hope I got pnpm-specific terminology right there) can not have their own dependencies? Just guessing here.

Meaning, since now `my-project/package.json` has only shared dependencies of all 3 sub-projects ( `components/backend`, `components/frontend`, and `components/protocol`), while each of those sub-projects has further dependencies in their `package.json`?

**Comment by AndreasNasman** [comment]
Hmm, yeah could be.

`backend/node_modules/@ty-ras` contains a `backend-node-io-ts-openapi` symlink to the root `node_modules`, but the path looks a bit odd going first to `@ty-ras+backend-node-io-ts-openapi@0.13.5` and nesting further `@ty-ras/backend-node-io-ts-openapi` 🤔

`backend-node-io-ts-openapi -> ../../../../node_modules/.pnpm/@ty-ras+backend-node-io-ts-openapi@0.13.5/node_modules/@ty-ras/backend-node-io-ts-openapi`
![image](https://user-images.githubusercontent.com/17124245/236697514-dea67b2d-da54-4806-a68a-631d99548d24.png)
![image](https://user-images.githubusercontent.com/17124245/236697654-e92cc005-240f-454e-a580-f3416a579a8a.png)
![image](https://user-images.githubusercontent.com/17124245/236697679-cb90a611-4269-4012-a548-9190f8e77305.png)
![image](https://user-images.githubusercontent.com/17124245/236697709-8ace7d12-b24e-4797-abf7-5632a25d6fcc.png)
![image](https://user-images.githubusercontent.com/17124245/236697751-0223379f-e056-4e12-9396-e2649286087b.png)

**Comment by AndreasNasman** [comment]
Doing [pnpm install --shamefully-hoist=true](https://pnpm.io/it/npmrc#shamefully-hoist) resolves this issue. I'll have a google and see if some "proper" fix is hard to apply 😄

**Comment by stazz** [comment]
I love how opinionated that CLI argument name is 😂

**Comment by AndreasNasman** [comment]
The proper fix seems to be to include `openapi-types` as a dependency in `backend/package.json`. That solves the problem and `pnpm install` can be run as normal 😊

---

The idea of pnpm is to require every `import`ed package explicitly listed in `package.json` to avoid "silly bugs":
https://www.kochan.io/nodejs/pnpms-strictness-helps-to-avoid-silly-bugs.html

I actually run into this exact "silly bug" problem a while ago with Jest when using `@jest/globals` implicitly, i.e. not having it listed explicitly in `package.json`. The implicit version was outdated so stuff wasn't working as expected, but it took a while to figure it out since their documentation was missing a note on it 😅

**Comment by stazz** [comment]
I don't really agree with `pnpm`'s idea of this, as it results in overly verbose and excessive `package.json` files which no one really cares to investigate in case of something goes wrong. But I will make a fix so that for `pnpm` all the `import`ed packages will be included as explicit dependencies. 👍  Thanks for digging into this!

**Comment by stazz** [comment]
This is now fixed in version `0.13.5`. I added a testcase for this scenario, and then made a fix. The fix was, instead of adding explicit dependency, to modify all templates to use `tyras.FinalMetadata` instead of `openapi.Document` type.
